src/fluid.rs

```diff
@@ -0,0 +1,434 @@
+extern crate image;
+extern crate rand;
+
+use geo::algorithm::rotate::RotatePoint;
+use geo::{line_string, point};
+use noise::{NoiseFn, Perlin};
+use rand::Rng;
+
+//#[derive(Copy, Clone)]
+pub struct SimulationConfig {
+    pub dt: f32,
+    pub iter: u32,
+    pub scale: u32,
+    pub t: f32,
+    pub size: u32,
+}
+
+impl Default for SimulationConfig {
+    fn default() -> SimulationConfig {
+        SimulationConfig {
+            dt: 0.02,
+            iter: 16,
+            scale: 4,
+            t: 0f32,
+            size: 128,
+        }
+    }
+}
+
+//#[derive(Copy, Clone)]
+pub struct FluidConfig {
+    pub diffusion: f32,
+    pub viscousity: f32,
+}
+
+impl Default for FluidConfig {
+    fn default() -> FluidConfig {
+        FluidConfig {
+            diffusion: 0f32,
+            viscousity: 0.0000001,
+        }
+    }
+}
+
+//#[derive(Copy, Clone)]
+pub struct Fluid {
+    pub fluid_configs: FluidConfig,
+    pub simulation_configs: SimulationConfig,
+    s: Vec<f32>,
+    density: Vec<f32>,
+    velocities_x: Vec<f32>,
+    velocities_y: Vec<f32>,
+    velocities_x0: Vec<f32>,
+    velocities_y0: Vec<f32>,
+}
+
+impl Fluid {
+    pub fn new(init_fluid: FluidConfig, init_simulation: SimulationConfig) -> Fluid {
+        Fluid {
+            s: vec![0.0; (init_simulation.size * init_simulation.size) as usize],
+            density: vec![0.0; (init_simulation.size * init_simulation.size) as usize],
+            velocities_x: vec![0.0; (init_simulation.size * init_simulation.size) as usize],
+            velocities_y: vec![0.0; (init_simulation.size * init_simulation.size) as usize],
+            velocities_x0: vec![0.0; (init_simulation.size * init_simulation.size) as usize],
+            velocities_y0: vec![0.0; (init_simulation.size * init_simulation.size) as usize],
+            fluid_configs: init_fluid,
+            simulation_configs: init_simulation,
+        }
+    }
+
+    /// Returns the variable's value constrained between from and to (both inclusive)
+    /// # Arguments
+    /// * `var` - The variable whoose value to be processed
+    /// * `from` - the minimal value that can be hold (inclusive)
+    /// * `to` - the maximum value that can be hold (inclusive)
+    fn constrain<T: PartialOrd>(var: T, from: T, to: T) -> T {
+        match var {
+            d if d < from => from,
+            d if d > to => to,
+            _ => var,
+        }
+    }
+
+    fn coordinate_to_idx(x: u32, y: u32, size: &u32) -> usize {
```

Comment:
This could be probably made with marco_rules! and named shorter, something like idx!(x, y) to be easier to write & read later

Reply:
this is little verbose for something that is called non stop. the function is small enough to be optimized and inlined, so there should be no difference in speed, it will be just easier to write. Again you decide how it should be